From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maksim Korotkov <m.korotkov@postgrespro.ru>
Date: Mon, 13 Jul 2026 13:16:40 +0300
Subject: [PATCH 3/4] address_standardizer: clean up partial 2D allocations

Ported-from: https://github.com/postgis/address_standardizer/commit/330d63a4723f9506c6a57ae10c595626db7f991d
(cherry picked from commit c38c8731915bda57a5baf1feb3a24ac393eade7e)
---
 extensions/address_standardizer/pagc_api.h | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/extensions/address_standardizer/pagc_api.h b/extensions/address_standardizer/pagc_api.h
index 7267db047..d26bee482 100644
--- a/extensions/address_standardizer/pagc_api.h
+++ b/extensions/address_standardizer/pagc_api.h
@@ -1095,7 +1095,16 @@ int tokenize_landmark_words(char *, char **) ;
       int row_num ; \
       PAGC_CALLOC_STRUC(temp_ptr,TYP*,ROWS,WHERE,EXIT_TYPE) ; \
       for ( row_num = 0 ; row_num < ROWS ; row_num++ ) { \
-        PAGC_CALLOC_STRUC(temp_ptr[row_num],TYP,COLS,WHERE,EXIT_TYPE) ; \
+        /* Manual calloc: PAGC_CALLOC_STRUC can't clean up prior rows */ \
+        temp_ptr[row_num] = (TYP*)calloc(COLS, sizeof(TYP)); \
+        if (temp_ptr[row_num] == NULL) { \
+          while (row_num) { \
+            row_num--; \
+            free(temp_ptr[row_num]); \
+          } \
+          FREE_AND_NULL(temp_ptr); \
+          RET_ERR("Insufficient Memory", WHERE, EXIT_TYPE); \
+        } \
       } \
       PTR = temp_ptr ; \
    }
